sklearn/metrics/pairwise.py

```diff
@@ -1130,6 +1220,227 @@ def _pairwise_callable(X, Y, metric, **kwds):
+                                           block_n_rows=1, **kwds):
+    """Generates blocks of the distance matrix from X and optional Y.
+
+    Parameters
+    ----------
+    X : array [n_samples_a, n_samples_a] if metric == "precomputed", or,
+        [n_samples_a, n_features] otherwise
+        Array of pairwise distances between samples, or a feature array.
+
+    Y : array [n_samples_b, n_features], optional
+        An optional second feature array. Only allowed if
+        metric != "precomputed".
+
+    metric : string, or callable
+        The metric to use when calculating distance between instances in a
+        feature array.
+
+    n_jobs : int
+        The number of jobs to use for the computation.
+
+    block_size : int, default=64
+        The maximum number of mebibytes (MiB) of memory per job (see``n_jobs``)
+        to use at a time for calculating pairwise distances.
+
+    block_n_rows : int
+        Number of rows to be computed for each block.
+
+    `**kwds` : optional keyword parameters
+        Any further parameters are passed directly to the distance function.
+
+    Returns
+    -------
+    D : generator of blocks based on the ``block_size`` parameter.
+
+    """
+    if metric != 'precomputed' and Y is None:
+        Y = X
+    n_samples = X.shape[0]
+    for start in range(0, n_samples, block_n_rows):
+        # get distances from block to every other sample
+        stop = min(start + block_n_rows, X.shape[0])
+        yield pairwise_distances(X[start:stop], Y, metric, n_jobs, **kwds)
+
+
+def pairwise_distances_reduce(X, Y=None, reduce_func=None, metric='euclidean',
+                              n_jobs=1, block_size=DEFAULT_BLOCK_SIZE, **kwds):
+    """Compute the distance matrix from a vector array X and optional Y.
+
+    This method takes either a vector array or a distance matrix, and a
+    reducing function to reduce each block of the distance matrix produced,
+    as per the block_size parameter. If the input is a vector array, the
+    distances are computed. If the input is a distances matrix, it is reduced
+    in size and returned instead.
+
+    This is equivalent to calling:
+
+        pairwise_distances(X, y, metric, n_jobs)
+
+    but should use less memory.
+
+    Parameters
+    ----------
+    X : array [n_samples_a, n_samples_a] if metric == "precomputed", or,
+        [n_samples_a, n_features] otherwise
+        Array of pairwise distances between samples, or a feature array.
+
+    Y : array [n_samples_b, n_features], optional
+        An optional second feature array. Only allowed if
+        metric != "precomputed".
+
+    reduce_func : function, callable
+        The function which is applied on each block of the distance matrix
+        reducing its size. It reduces the size of each block from
+        [n_block_samples, n_samples_a] or [n_block_samples, n_samples] to
+        [n_block_samples, n_reduced] where n_block_samples is the number of
```

Comment:
Actually it can be any shape beginning `n_block_samples`, as long as the remaining dimensions are consistent across calls to `reduce_func`

Note that with `flexible_concatenate`, `reduce_func` can return a tuple of such reductions. Also note that the return value can be array-like or sparse matrix.